ecosystem/sep-0039.md

```diff
@@ -0,0 +1,182 @@
+
+```python
+import string 
+ALPHABET = string.digits + string.ascii_lowercase
+
+def int_to_b36(i: int) -> str:
+    d, r = divmod(i, len(ALPHABET))
+    return ALPHABET[d] + ALPHABET[r]
+
+def b36_to_int(s: str) -> int:
+    return ALPHABET.index(s[0]) * len(ALPHABET) + ALPHABET.index(s[1])
+```
+
+Another consideration is perhaps trying to have **two** indices within this value. Since we need 10 bits, we can use the other 6 for identifying this asset within the account itself. There's still the question of printables, but  though, so this needs more thought.
+
+---------
+
+#### Version
+The first entry's index is followed by the version string. For the current version of this SEP, this is just `1`.
+
+#### Metadata
+The **metadata** is a [Media Type](https://www.iana.org/assignments/media-types/media-types.xhtml) alongside any associated parameters. It is prefixed by a **length**, which is just a stringified integer (that is, a 32-byte metadata will have "32" as its length). When parsing, you should consume characters until hitting a non-digit, since a valid Media Type will never start with a digit. Note that row indices are not included in the metadata length.
+
+**Author's note**: If we go with a radix of 36, we could use that here, too.
+
+There are no *required* parameters.
+
+Standard *optional* parameters include:
+
+  * `n`: this defines the name or "title" of your asset
+  * `d`: this provides a description your asset
+  * `c`: a crc32 checksum of the asset
+
+Metadata _must_ only be composed of printable ASCII characters.
+
+#### Binary Data
+With the header out of the way, we need to describe the way the remaining binary data should be represented in `ManageData` entries.
+
+Since the _key_ is restricted specifically to printable characters (ref: Stellar Core's [`isString32Valid()`](https://github.com/stellar/stellar-core/blob/b031954458df3bb31d8d98f136c6fee40523a10d/src/util/types.cpp#L89-L101) and C's [`isprint()`](https://en.cppreference.com/w/cpp/string/byte/isprint)), this gives us an encoding space of 95 characters or ~6.5 bits per character (a ~25% reduction in the space). From our research, the [BasE91] encoding scheme is closest to this theoretical limit.
+
+The _value_ can be any binary data; thus, no encoding is necessary.
+
+All of the binary data is thus prepended with an index, the next BasE91-encoded chunk sized to fit into the 64-length key string, and the next 64-byte binary chunk as the value.
+
+If the final entry fits into the key (that is, the value is not necessary), the encoder _must_ set the value to an empty buffer (as opposed to `nil` or `null`). Otherwise, this will delete the entry.
+
+#### Implementation Note
```

Comment:
I'm confused here. What is this solving for. If we can submit empty buffers as values we just keep chunking along until we run out of actual chunks and we finish with an empty buffer value.

Reply:
That's exactly what it's saying here: if you're out of data on the last row for the value, you still need set it to be an empty buffer. It's exactly what you said [here](https://github.com/stellar/stellar-protocol/pull/1090#discussion_r759331013).

Reply:
Sorry I meant the `Implementation Note`
> Because you cannot predict the length of your encoded string without doing the encoding itself, it will be necessary to use trial-and-error to find the longest binary chunk that fits in <= 64 bytes. While this does mildly impact encoding time, it ensures optimality on-chain.

Reply:
Oh, this is actually related to your [later comment](https://github.com/stellar/stellar-protocol/pull/1090#issuecomment-984858571): base64 will always have 48 -> 64 because of padding, but base91 is less predictable (part of the reason why it's more efficient). That means you need to "guess and check" until you max out the buffer.

Reply:
It's impossible for base91 to do better than 10% overhead, so if you search from `ceil(n/1.10)` (where `n` is the final length you _want_), you'll get it. This was also defined in `_encode_nearest()` in the original commit (see [here](https://github.com/stellar/stellar-protocol/pull/1090/commits/9668dbe582319e0e83c5f427e028dee454411a43#diff-9e64d12ca10dd7b12e3632c74b968e5d35f418587ee64f6d92fb0d9aac347c65R136-R147)), if you want a reference implementation.